### PR TITLE
Replace webpack.optimize.UglifyJsPlugin to UglifyJsPlugin

### LINKED ISSUE #351: Issues with Webpack 4
Hi!

There is a problem with the [recently released](https://github.com/webpack/webpack/releases) Webpack 4. It seems to be an issue with this plugin ussing `UglifyJsPlugin`, that now is deprecated. The following error is shown:

```
webpack.optimize.UglifyJsPlugin has been removed, please use config.optimization.minimize instead.
```

If I remove offline-plugin, then this error isn't there anymore.

### PR DESCRIPTION
**Pull request by ykzts**
Resolve #351

Drop support for Node.js v4 and webpack v1.

### CODE REVIEW

**Comment by NekR** [comment]
@ykzts wow, amazing! I'll proper review later today, thank you 👍

File: tests/browser/cache-first/webpack.config.js [hunk 1]
```diff
@@ -3,9 +3,11 @@ const { resolve, join } = path;
 
 const OfflinePlugin = require(process.env.OFFLINE_PLUGIN_ROOT);
 const CopyWebpackPlugin = require('copy-webpack-plugin');
-const DefinePlugin = require('webpack').DefinePlugin;
+const DefinePlugin = require('webpack/lib/DefinePlugin');
```
Comment (NekR) [inline tests/browser/cache-first/webpack.config.js]: This works for webpack 2+, right?

File: tests/browser/cache-first/www/main.js [hunk 2]
```diff
@@ -1,112 +1,141 @@
...
+var appCacheIframe;
 
-	function hasSW() {
-	  return 'serviceWorker' in navigator &&
-	    // This is how I block Chrome 40 and detect Chrome 41, because first has
-	    // bugs with history.pustState and/or hashchange
-	    (window.fetch || 'imageRendering' in document.documentElement.style) &&
-	    (window.location.protocol === 'https:' || window.location.hostname === 'localhost' || window.location.hostname.indexOf('127.') === 0)
-	}
+function hasSW() {
```
Comment (NekR) [inline tests/browser/cache-first/www/main.js]: So generated bundle output has been changed from webpack 1 to webpack 2, I remember that. Is it kept the same from 2 to 4? Judging by passed tests I guess it is 😃

**Comment by NekR** [comment]
Just for a note: Travis is failing because of broken puppeteer tests, i.e. not related to this PR.

**Comment by GGAlanSmithee** [comment]
@ykzts this is super awesome, thanks for submitting 🎉 

@NekR thanks for cc:ing :+1: 

Yes, it's def. a major version release since it drops support for webpack 1 which is a non backwards compatible change to the public API.

Also, I am totally fine with dropping support for AppCache in version 5, and i'd help you updating the docs if you want me too.

Here is a support table comparison for appcache vs. service worker:

[AppCache](https://caniuse.com/#feat=offline-apps)
[Service Worker](https://caniuse.com/#feat=serviceworkers)

So, depending on when version 5 is released, it looks like IE11 is the only browser which would loose support, since Safari and iOS Safari will support service worker in their next respective release.

(Sorry for derailing the thread 🙇)

**Comment by NekR** [comment]
@GGAlanSmithee Yeah, I think you're right. I actually meant just to disable AppCache by default, not drop its support, yet. So it should be even safer (i.e. with a way to turn it on again) :-)

File: tests/browser/cache-first/webpack.config.js (hunk 1 above)
Comment (ykzts) [reply tests/browser/cache-first/webpack.config.js]: - [v4.1.0](https://github.com/webpack/webpack/blob/v4.1.0/lib/DefinePlugin.js)
- [v3.11.0](https://github.com/webpack/webpack/blob/v3.11.0/lib/DefinePlugin.js)
- [v2.7.0](https://github.com/webpack/webpack/blob/v2.7.0/lib/DefinePlugin.js)

👍

**Comment by ykzts** [comment]
> Just for a note: Travis is failing because of broken puppeteer tests, i.e. not related to this PR.

I added `args: ['--no-sandbox', '--disable-setuid-sandbox']` with reference to https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md. CI succeeded 😄

File: tests/browser/cache-first/www/main.js (hunk 2 above)
Comment (ykzts) [reply tests/browser/cache-first/www/main.js]: In webpack v2 to v4, it seems to be different compilation results... Maybe it should be excluded with .gitignore.

**Comment by NekR** [comment]
Merged, thank you! Making a v5 release soon 🎉

**Comment by audiolion** [comment]
fantastic news! thanks @ykzts

**Comment by audiolion** [comment]
@NekR could you release a `@next` version, a `5.0.0.alpha-0` or similar so that we can get the webpack v4 support?

**Comment by NekR** [comment]
@audiolion Yeah, indeed, makes sense. Here you go:

```
> npm publish --tag=next

+ offline-plugin@5.0.0-0
```

**Pull request merged**